classes/oai/ojs/OAIDAO.inc.php

```diff
@@ -252,72 +242,90 @@ public function _getRecordsRecordSet($setIds, $from, $until, $set, $submissionId
+                's.section_id AS section_id',
+            ])
+            ->join('publications AS p', 'a.current_publication_id', '=', 'p.publication_id')
+            ->join('publication_settings AS psissue', function($join) {
+                $join->on('psissue.publication_id', '=', 'p.publication_id');
+                $join->where('psissue.setting_name', '=', DB::raw('\'issueId\''));
+                $join->where('psissue.locale', '=', DB::raw('\'\''));
+            })
+            ->join('issues AS i', DB::raw('CAST(i.issue_id AS CHAR(20))'), '=', 'psissue.setting_value')
+            ->join('sections AS s', 's.section_id', '=', 'p.section_id')
+            ->join('journals AS j', 'j.journal_id', '=', 'a.context_id')
+            ->where('i.published', '=', 1)
+            ->where('j.enabled', '=', 1)
+            ->where('a.status', '=', PKPSubmission::STATUS_PUBLISHED)
+            ->when($excludeJournals, function($query, $excludeJournals) {
+                return $query->whereNotIn('j.journal_id', $excludeJournals);
+            })
+            ->when(isset($journalId), function($query) use ($journalId) {
+                return $query->where('j.journal_id', '=', (int) $journalId);
+            })
+            ->when(isset($sectionId), function($query) use ($sectionId) {
+                return $query->where('p.section_id', '=', (int) $sectionId);
+            })
+            ->when($from, function($query, $from) {
+                return $query->where('GREATEST(a.last_modified, i.last_modified)', '>=', $from);
+            })
+            ->when($until, function($query, $until) {
+                return $query->where('GREATEST(a.last_modified, i.last_modified)', '<=', $until);
+            })
+            ->when($submissionId, function($query, $submissionId) {
+                return $query->where('a.submission_id', '=', (int) $submissionId);
+            })
+            ->union(
+                DB::table('data_object_tombstones AS dot')
+                ->select([
+                        'dot.date_deleted AS last_modified',
+                        'dot.data_object_id AS submission_id',
+                        DB::raw('NULL AS issue_id'),
+                        'dot.tombstone_id',
+                        'dot.set_spec',
+                        'dot.oai_identifier'
+                ])
+                ->when(isset($journalId), function($query, $journalId) {
+                    return $query->join('data_object_tombstone_oai_set_objects AS tsoj', function($join) use ($journalId) {
+                          $join->on('tsoj.tombstone_id', '=', 'dot.tombstone_id');
+                          $join->where('tsoj.assoc_type', '=', ASSOC_TYPE_JOURNAL);
+                          $join->where('tsoj.assoc_id', '=', (int) $journalId);
+                        })->addSelect(['tsoj.assoc_id']);
+                    }, function($query) {
+                        return $query->addSelect([DB::raw('NULL AS assoc_id')]);
+                })
```

Comment:
Why are the 4 last lines (addSelects) needed?

Reply:
The column sets in the two halves of the UNION statement need to be consistent and ordered the same, i.e. the `a, b, c, d` and `c, d, e, f` in
```
SELECT
  a, b, c, d
  FROM ...
UNION SELECT
  c, d, e, f
  FROM ...
```
This keeps the column set consistent with the other half of the `UNION`.